qoffer-infrastructure/src/main/groovy/life/qbic/portal/qoffer2/web/views/create/offer/SelectItemsView.groovy

```diff
@@ -75,53 +73,59 @@ class SelectItemsView extends VerticalLayout{
         this.createOfferViewModel = createOfferViewModel
         this.viewModel = viewModel
 
-        //todo translate the fetched products into the ProductItemViewModel objects to handle them in the gui
-        foundProducts = createOfferViewModel.foundProducts as ObservableList
-        foundProducts.addPropertyChangeListener(listener)
-
-        this.sequencingProduct = []
-        this.projectManagementProduct = []
-        this.storageProduct = []
-        this.primaryAnalyseProduct = []
-        this.secondaryAnalyseProduct = []
-        this.selectedItems = []
-
-        this.createOfferViewModel.productItems = selectedItems
-
-        initLayout()
-        setupDataProvider()
-        addListener()
-    }
-
-    /**
-     * Listens for changes of the foundProducts list and creates ProductItemViewModel objects from them
-     */
-    def listener = {
-        if (it instanceof ObservableList.ElementEvent)  {
-            foundProducts.each {
-                if(it instanceof Sequencing){
-                    sequencingProduct.add(new ProductItemViewModel(0, it))
+        sequencingProduct = createOfferViewModel.sequencingProducts as ObservableList
+        sequencingProduct.addPropertyChangeListener({
+            if (it instanceof ObservableList.ElementEvent) {
+                sequencingProduct.each {
                     sequencingGrid.dataProvider.refreshAll()
                 }
-                else if(it instanceof ProjectManagement){
-                    projectManagementProduct.add(new ProductItemViewModel(0, it))
+            }
+        })
+
+        projectManagementProduct = createOfferViewModel.managementProducts as ObservableList
+        projectManagementProduct.addPropertyChangeListener({
+            if (it instanceof ObservableList.ElementEvent) {
+                projectManagementProduct.each {
                     projectManagementGrid.dataProvider.refreshAll()
                 }
-                else if(it instanceof PrimaryAnalysis){
-                    primaryAnalyseProduct.add(new ProductItemViewModel(0, it))
+            }
+        })
+
+        storageProduct = createOfferViewModel.storageProducts as ObservableList
+        storageProduct.addPropertyChangeListener({
+            if (it instanceof ObservableList.ElementEvent) {
+                storageProduct.each {
+                    storageGrid.dataProvider.refreshAll()
+                }
+            }
```

Comment:
You can remove the list iteration here also. The data provider knows which list it is getting data from and you only need to refresh all once.

Reply:
700e6e4